### ISSUE TITLE
Clusters with Calico CNI generates every 5 minutes a calico-ctrl-change-..... configmaps in tkg-system namespace

### ISSUE THREAD

**Issue by trantor1** [vmware-tanzu/community-edition#4182]
## Bug Report

I'm observing a strange issue with the Calico CNI package in workload clusters.

A new config map is generated about every 5 minutes in tkg-system namesspace.

## Screenshots or additional information and context

``` sh
kubectl get cm -n tkg-system --sort-by=.metadata.creationTimestamp

calico-ctrl-change-49d2h                 1      175m
calico-ctrl-change-9h556                 1      169m
calico-ctrl-change-pn7wd                 1      164m
calico-ctrl-change-llkdm                 1      159m
calico-ctrl-change-zbqlh                 1      153m
calico-ctrl-change-8xx8r                 1      148m
calico-ctrl-change-w44r6                 1      143m
calico-ctrl-change-dj42q                 1      137m
calico-ctrl-change-h9n99                 1      132m
calico-ctrl-change-qs42n                 1      127m
calico-ctrl-change-q5ffp                 1      122m
calico-ctrl-change-gfgwx                 1      116m
calico-ctrl-change-bzq8q                 1      111m
calico-ctrl-change-gmfn4                 1      106m
calico-ctrl-change-6b2xr                 1      100m
calico-ctrl-change-xf5rg                 1      95m
calico-ctrl-change-c7c9r                 1      90m
calico-ctrl-change-m4trs                 1      84m
calico-ctrl-change-sz8dt                 1      79m
calico-ctrl-change-d5fzp                 1      74m
calico-ctrl-change-l7vnv                 1      69m
calico-ctrl-change-ltlbr                 1      63m
calico-ctrl-change-bhjdh                 1      58m
calico-ctrl-change-dvrgq                 1      53m
calico-ctrl-change-hg8dn                 1      47m
calico-ctrl-change-fvfh4                 1      42m
calico-ctrl-change-xr7t2                 1      37m
calico-ctrl-change-l2kg4                 1      31m
calico-ctrl-change-9jj2l                 1      26m
calico-ctrl-change-cfv8g                 1      21m
calico-ctrl-change-dxffj                 1      15m
calico-ctrl-change-xn5p9                 1      10m
calico-ctrl-change-cn4nl                 1      5m22s
calico-ctrl-change-ndpsw                 1      4s
``` 

* Installed Calico Version
``` 
$ tanzu package installed list -n tkg-system
- Retrieving installed packages... 
  NAME                  PACKAGE-NAME                           PACKAGE-VERSION        STATUS               
  calico                calico.tanzu.vmware.com                3.19.1+vmware.1-tkg.3  Reconcile succeeded  
```
* example of 2 last configmaps
```yaml
apiVersion: v1
data:
  spec: '{"startedAt":"2022-04-25T18:44:17.594402417Z","finishedAt":"2022-04-25T18:44:18.641692255Z","successful":true,"description":"update:
    Op: 0 create, 0 delete, 0 update, 15 noop / Wait to: 15 reconcile, 0 delete, 0
    noop","namespaces":["(cluster)","kube-system"]}'
kind: ConfigMap
metadata:
  creationTimestamp: "2022-04-25T18:44:17Z"
  generateName: calico-ctrl-change-
  labels:
    kapp.k14s.io/app-change-app: calico-ctrl
    kapp.k14s.io/is-app-change: ""
  name: calico-ctrl-change-lwtpg
  namespace: tkg-system
  resourceVersion: "292217"
  uid: c0050c6b-88e4-4f11-8b1f-2a10b0f9144a
---
apiVersion: v1
data:
  spec: '{"startedAt":"2022-04-25T18:49:37.513563411Z","finishedAt":"2022-04-25T18:49:38.556650676Z","successful":true,"description":"update:
    Op: 0 create, 0 delete, 0 update, 15 noop / Wait to: 15 reconcile, 0 delete, 0
    noop","namespaces":["(cluster)","kube-system"]}'
kind: ConfigMap
metadata:
  creationTimestamp: "2022-04-25T18:49:37Z"
  generateName: calico-ctrl-change-
  labels:
    kapp.k14s.io/app-change-app: calico-ctrl
    kapp.k14s.io/is-app-change: ""
  name: calico-ctrl-change-29tpr
  namespace: tkg-system
  resourceVersion: "293343"
  uid: 7c5c8517-3862-4b08-923f-1433b623258c
```

* calico-ctrl self
``` yaml
apiVersion: v1
data:
  spec: '{"labelKey":"kapp.k14s.io/app","labelValue":"1650830493043029977","lastChangeName":"calico-ctrl-change-29tpr","lastChange":{"startedAt":"2022-04-25T18:49:37.513563411Z","finishedAt":"2022-04-25T18:49:38.556650676Z","successful":true,"description":"update:
    Op: 0 create, 0 delete, 0 update, 15 noop / Wait to: 15 reconcile, 0 delete, 0
    noop","namespaces":["(cluster)","kube-system"]},"usedGVs":[{"Group":"","Version":"v1"},{"Group":"apiextensions.k8s.io","Version":"v1"},{"Group":"rbac.authorization.k8s.io","Version":"v1"},{"Group":"apps","Version":"v1"},{"Group":"policy","Version":"v1beta1"}]}'
kind: ConfigMap
metadata:
  creationTimestamp: "2022-04-24T20:01:33Z"
  labels:
    kapp.k14s.io/is-app: ""
  name: calico-ctrl
  namespace: tkg-system
  resourceVersion: "293344"
  uid: 2a193350-b938-4c37-868d-d7478eb206f9
```

## Environment Details

* Build version (`tanzu version`): 0.11.0
* Deployment (Managed/Unmanaged cluster): managed
* Infrastructure Provider (Docker/AWS/Azure/vSphere): vSphere
* Operating System (client): linux

**Comment by stmcginnis** [comment]
@12345lcr can you take a look?

**Comment by 12345lcr** [comment]
Nice finding. I took some time to look into this and found that this is because kapp unnecessarily creates change configmap in kapp controller even when there is no change. Corresponding kapp [issue#395](https://github.com/vmware-tanzu/carvel-kapp/issues/395)

### Root Cause
The root cause is we defined status in calico CRD, and kapp reads this as set status to following value which would reset what was populated in the cluster.
```
status:
  acceptedNames:
    kind: ""
    plural: ""
  conditions: []
  storedVersions: []
```
For every reconciliation(we set sync period to 5mins), this reset happens and will create a ctrl-change configmap. 
And kapp defaults to maximum 200 ctrl-change configmap records for each package to keep the history.

### Workaround
The workaround is to delete the status fields from CRDs. Similar workaround has already applied to ako-operator and load-balancer-and-ingress-service.
Ako-operator: https://github.com/vmware-tanzu/community-edition/commit/71afbaea4551af1afc24decd05669614b200e020
load-balancer-and-ingress-service: https://github.com/vmware-tanzu/community-edition/pull/3305/commits/8eed3d5293b8a44a98d70536ea67faa157694e43

### Fix
The fix in kapp is already merged and released starting from v0.45.0
https://github.com/vmware-tanzu/carvel-kapp/pull/411

**Comment by 12345lcr** [comment]
This issue should be fixed after kapp-controller is bumped to 0.34.0+ in next release.
 
kapp-controller 0.34.0 is using kapp v0.46.0, which contains the fix in v0.45.0 that mentioned in the issue description.
https://github.com/vmware-tanzu/carvel-kapp-controller/blob/v0.34.0/hack/dependencies.yml#L8
```
kapp_version: v0.46.0
```
